internal/dataplane/parser/parser.go

```diff
@@ -386,14 +390,134 @@ func getCertFromSecret(secret *corev1.Secret) (string, string, error) {
+							// TODO support cert_alt and key_alt if there are 2 SecretObjectReferences
+							// https://github.com/Kong/kubernetes-ingress-controller/issues/2604
+							log.WithFields(logrus.Fields{
+								"gateway":  g.Name,
+								"listener": l.Name,
+							}).Error("Gateway Listeners with more than one certificateRef are not supported")
+							continue
+						}
+						namespace := g.Namespace
+						ref := l.TLS.CertificateRefs[0]
+						// SecretObjectReference is a misnomer; it can reference any Group+Kind, but defaults to Secret
+						if ref.Kind != nil {
+							if string(*ref.Kind) != "Secret" {
+								log.WithFields(logrus.Fields{
+									"gateway":  g.Name,
+									"listener": l.Name,
+								}).Error("CertificateRefs to kinds other than Secret are not supported")
+							}
+						}
+						if ref.Namespace != nil {
+							namespace = string(*ref.Namespace)
+						}
+						if namespace != g.Namespace {
+							allowed := getPermittedForReferencePolicyFrom(gatewayv1alpha2.ReferencePolicyFrom{
+								Group:     gatewayv1alpha2.Group(g.GetObjectKind().GroupVersionKind().Group),
+								Kind:      gatewayv1alpha2.Kind(g.GetObjectKind().GroupVersionKind().Kind),
+								Namespace: gatewayv1alpha2.Namespace(g.GetNamespace()),
+							}, policies)
+							if !isRefAllowedByPolicy(ref.Namespace, ref.Name, ref.Group, ref.Kind, allowed) {
+								log.WithFields(logrus.Fields{
+									"gateway":           g.Name,
+									"gateway_namespace": g.Namespace,
+									"listener":          l.Name,
+									"secret_name":       string(ref.Name),
+									"secret_namespace":  namespace,
+								}).WithError(err).Error("secret reference not allowed by ReferencePolicy")
+								continue
+							}
+						}
+						secret, err := s.GetSecret(namespace, string(ref.Name))
+						if err != nil {
+							log.WithFields(logrus.Fields{
+								"gateway":          g.Name,
+								"listener":         l.Name,
+								"secret_name":      string(ref.Name),
+								"secret_namespace": namespace,
+							}).WithError(err).Error("failed to fetch secret")
+							continue
+						}
+						cert, key, err := getCertFromSecret(secret)
+						if err != nil {
+							log.WithFields(logrus.Fields{
+								"gateway":          g.Name,
+								"listener":         l.Name,
+								"secret_name":      string(ref.Name),
+								"secret_namespace": namespace,
+							}).WithError(err).Error("failed to construct certificate from secret")
+							continue
+						}
+						hostname := "*"
+						if l.Hostname != nil {
+							hostname = string(*l.Hostname)
+						}
+						certs = append(certs, certWrapper{
+							identifier: cert + key,
+							cert: kong.Certificate{
+								ID:   kong.String(string(secret.UID)),
+								Cert: kong.String(cert),
+								Key:  kong.String(key),
+							},
+							CreationTimestamp: secret.CreationTimestamp,
+							snis:              []string{hostname},
+						})
+					}
+				}
+			}
+		}
+	}
+	return certs
+}
```

Comment:
Aside from the Condition check, what are you looking to break out here? As I read this, the actual actions themselves aren't particularly complex, it's just longish because we have a bunch of multi-line struct definitions, mostly for the log fields.

Reply:
I suppose what I was thinking about most was seeing if we could tighten up those log lines yes. However, this doesn't need to be a blocker.